buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java

```diff
@@ -173,15 +174,29 @@ public String getName() {
 
     @Internal
     public Version getVersion() {
-        return distribution.getVersion();
+        return distributions.get(currentDistro).getVersion();
     }
 
     @Override
     public void setVersion(String version) {
         requireNonNull(version, "null version passed when configuring test cluster `" + this + "`");
+        setVersion(Collections.singletonList(version));
+    }
+
+    @Override
+    public void setVersion(List<String> versions) {
```

Comment:
We really don't have a need to go through multiple versions. The upgrade tests start with one version, and go to another. This seems far more complicated to me than we had before, where each step of the upgrade tests was a different cluster. IIRC the tricky part there was hooking up finalizers to shut down certain nodes between phases of the tests, but given that testclusters has far better control over when nodes start and stop, why couldn't we stick with that pattern?

Reply:
I'm completely open to suggestion here, that's why I stopped at converting one project.  Here's my reasoning  for going down this path. 

The reason we can't just call `setVersion` from a `doFirst` is that all versions need to be known at configuration time. 
 
I wanted to have a way to support the upgrade in the framework to have less repetition in the projects that implement bwc testing. 
The alternative would be to have multiple cluster definitions and set the data paths ( which we would have to allow, we don't currently ). 
For rolling restart we would have to come up with a way to tell testclusters not to start the cluster for the task, as the task would want to have fine grained control and start the upgraded cluster node-by-node, I really dislike having to add that.

I would also prefer we don't give node level control to build scripts as it would make  implementing the tests more imperative, but descriptive tests are easier to understand. 

Another use-case that was brought up and plays into this is testing with evolving configurations in a mixed cluster. e.x. starting with some nodes that have ML disabled, run some tests, enable them, run some more. 
In addition we already have some tests that alter the configuration of the cluster ( e.x. enabling security ) that currently just call the same configuration DSL and restart the cluster, but I'm not happy with them.  

The initial thinking with the `freeze()` call was that all configuration would happen at Gradle's configuration time, thus the versions and all the config should be set at Gradle's configuration time and frozen during execution time. 

The `setVersion(List)` is probably not the best way to express this, but I really think we should keep the paradigm of configuring both the initial and all future states of the cluster and then have a DSL that would move between the states.
I see `setVersion(List)` as a shortcut to get there and start running bwc with `--parallel` and get rid of all the timeout related test failures in CI before we go and refactor to separate out configuration from the implementation of the actions, so we can have something like named states we could go to, so we wouldn't need  `setVersions` but instead it would look more like configuring multiple clusters with configuration inheritance, except it won't be multiple clusters but different state of the same cluster.

Reply:
We discussed that we will merge this PR as is and discuss DSL in a follow up.